crates/biome_js_analyze/src/lint/nursery/no_substr.rs

```diff
@@ -0,0 +1,138 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::{JsCallExpression, JsSyntaxToken};
+use biome_rowan::{AstSeparatedList, BatchMutationExt, TextRange, TokenText};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Enforce the use of `slice()` over `substr()` and `substring()`.
+    ///
+    /// `slice()` is preferred over `substr()` and `substring()` because it is a more popular option with clearer behavior,
+    ///  and it has a consistent counterpart in arrays.
+    ///
+    /// Note that `substr`, `substring` and `slice` are not identical when arguments are passed.
+    /// For detailed differences, refer to the MDN documentation:
+    /// - [The difference between substring() and substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr)
+    /// - [Differences between substring() and slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#differences_between_substring_and_slice)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substr(start, length);
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substring(indexStart, indexEnd);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// foo.slice(beginIndex, endIndex);
+    /// ```
+    ///
+    pub NoSubstr {
+        version: "next",
+        name: "noSubstr",
+        language: "js",
+        recommended: false,
+        fix_kind: FixKind::Safe,
```

Comment:
The ix should be marked as unsafe.
```suggestion
        fix_kind: FixKind::Unsafe,
```